pkg/localenvoy/controllers/ingress/envoyingress.go

```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ingress
+
+import (
+	"context"
+	"fmt"
+	"hash/fnv"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+)
+
+const (
+	clusterLabel     = "kcp.dev/cluster"
+	toEnvoyLabel     = "ingress.kcp.dev/envoy"
+	ownedByCluster   = "ingress.kcp.dev/owned-by-cluster"
+	ownedByIngress   = "ingress.kcp.dev/owned-by-ingress"
+	ownedByNamespace = "ingress.kcp.dev/owned-by-namespace"
+)
+
+// reconcile is triggered on every change to an ingress resource, or it's associated services (by tracker).
+func (c *Controller) reconcile(ctx context.Context, ingress *networkingv1.Ingress) error {
+	klog.InfoS("reconciling Ingress", "ClusterName", ingress.ClusterName, "Namespace", ingress.Namespace, "Name", ingress.Name)
+
+	if ingress.Labels[clusterLabel] == "" {
+		// this is the root. Ignore.
```

Comment:
Am I interpreting the logic in here correctly?

- Only reconcile leaves
- Get the root
- Calculate the root LB hostname
- Label the leaf for Envoy (which will get updated in `process() after this function returns`)
- Update the root (to set LB hostname)

Would it make sense to split this into 1 part that reconciles a root (set hostname & update only if needed) and 1 part that reconciles a leaf?

Reply:
yes.  There's only one small thing, i've added a new "check" to make sure that the leaf has a status set, If it doesn't, we shouldn't mark it ready for envoy, and neither set the root Ingress status field, as we don't know if there's another ingress leaf with an status endpoint.

Also, perhaps we should avoid updating the root Ingress status if it doesn't change?